hep/proposals/0000-decorators.md

```diff
@@ -0,0 +1,299 @@
+    
+
+# This defines the another template input
+class FiboInput(hio.Input):
+    num: int
+
+class FiboOutput(hio.Output):
+    num: int
+
+@wt.entrypoint
+@wt.steps  # Adds a new script template to the workflow template called fibonacci
+def fibonacci(fibo: FiboInput) -> FiboOutput:
+    # we will need to validate that users can only run/define certain kinds of expressions here
+    # and throw warnings if they are using syntax that is invalid
+    valid_number = hio.expr(fibo.num > 1)
+    previous_num = calculator(CalculatorInput(x=fibo.num, y=1, operation="sub"), when=valid_number)
+    previous_fibo = fibonacci(FiboInput(num=previous_num.result), when=valid_number)
+    current = calculator(CalculatorInput(x=fibo.num, y=previous_fibo.num), when=valid_number)
+    return FiboOutput(num=valid_number.check(current.num, 1))
+```
+
+
+
+## Container Template
+
+```python
+from hera.workflows import WorkflowTemplate, Container
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class MyInput(hio.Input):
+    user: str = "Hera"
+
+class MyOutput(hio.Output):
+    container_greeting: Annotated[str, Parameter(name="container-greeting")]
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"], args=["echo {{input.parameters.user}}"])
+def basic_hello_world(my_input: MyInput) -> hio.Output:
+    ...
+
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"])
+def advanced_hello_world(my_input: MyInput, template: Container) -> MyOutput:
+    # hio.output serves two functions. One, it allows users to create the output class
+    # which allows output path variables to be referenced.
+    # Second, it can be used to "mock" a container template when running locally and allow us to
+    # inject outputs based on inputs.
+    output: MyOutput = hio.output(my_input)
+    # template is a special variable that allows you to reference the template itself and modify
+    # its attributes. This is especially useful when trying to reference input params in args
+    template.args = [f"echo {hio.name(my_input.user)} > {hio.path(output.container_greeting)}"]
+    return output
+```
+
+## Workflow Template Ref
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+from hera.expr import spring
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+ewt = WorkflowTemplate(name="externally-workflow-template")
+awt = WorkflowTemplate(name="another-workflow-template")
+
+@awt.script  # Adds a new script template to a workflow template called "another-workflow-template"
+def setup() -> hio.Output:
+    return hio.Output(result="Setting things up")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+
+class ConcatOutput(hio.Output):
+    value: str
+
+# Adds a new template ref from an external template
+# This can be auto-generated or typed by the user
+@ewt.template_ref
```

Comment:
Would `cluster_scope` go in the decorator args, or be derived from the class referenced (i.e. `WorkflowTemplate`/`ClusterWorkflowTemplate`)?

Reply:
Discussed and decided to remove the `template_ref` decorator from the HEP